Sources/SwiftProtobuf/AnyMessageStorage.swift

```diff
@@ -301,163 +323,173 @@ extension AnyMessageStorage {
-    // the conversion cost might be to high for those uses.  Give up and say they aren't equal.
-    return false
-  }
+      #if swift(>=4.2)
+        // Can't use _valueData for a few reasons:
+        // 1. Since decode is done on demand, two objects could be equal
+        //    but created differently (one from JSON, one for Message, etc.),
+        //    and the hash values have to be equal even if we don't have data
+        //    yet.
+        // 2. map<> serialization order is undefined. At the time of writing
+        //    the Swift, Objective-C, and Go runtimes all tend to have random
+        //    orders, so the messages could be identical, but in binary form
+        //    they could differ.
+        public func hash(into hasher: inout Hasher) {
+          if !_typeURL.isEmpty {
+            hasher.combine(_typeURL)
+          }
+        }
+      #else  // swift(>=4.2)
+        var hashValue: Int {
+          var hash: Int = i_2166136261
+          if !_typeURL.isEmpty {
+            hash = (hash &* i_16777619) ^ _typeURL.hashValue
+          }
+          return hash
+        }
+      #endif
+
+      func isEqualTo(other: AnyMessageStorage) -> Bool {
+            if _typeURL != other._typeURL { return false }
+
+            // Since the library does lazy Any decode, equality is a very hard problem.
+            // It things exactly match, that's pretty easy, otherwise, one ends up having
+            // to error on saying they aren't equal.
+            //
+            // The best option would be to have Message forms and compare those, as that
+            // removes issues like map<> serialization order, some other protocol buffer
+            // implementation details/bugs around serialized form order, etc.; but that
+            // would also greatly slow down equality tests.
+            //
+            // Do our best to compare what is present have...
+
+            // If both have messages, check if they are the same.
+            if case .message(let myMsg) = state,
+                  case .message(let otherMsg) = other.state, type(of: myMsg)
+                  == type(of: otherMsg) {
+                  // Since the messages are known to be same type, we can claim both equal and
+                  // not equal based on the equality comparison.
+                  return myMsg.isEqualTo(message: otherMsg)
+            }
+
+            // If both have serialized data, and they exactly match; the messages are equal.
+            // Because there could be map in the message, the fact that the data isn't the
+            // same doesn't always mean the messages aren't equal. Likewise, the binary could
+            // have been created by a library that doesn't order the fields, or the binary was
+            // created using the appending ability in of the binary format.
+            if case .binary(let myValue) = state,
+                  case .binary(let otherValue) = other.state, myValue
+                  == otherValue
```

Comment:
We should probably use a group around the individual clauses to encourage them to wrap as a full unit, and only break internally if needed beyond that.

Reply:
Mos' def!